WordPress/gutenberg
# List View: Experiment with drag in place

This PR explores alternative drag interactions using [motion framer](https://www.framer.com/docs/animation/). Instead of using a drag chip, we instead allow items to drag in place. Follow up to: https://github.com/WordPress/gutenberg/pull/33765

This PR is pretty large in size since I wasn't able to find any smaller chunks to safely land. If folks spot pieces we can break off, I'd be happy to break this into additional PRs.

https://user-images.githubusercontent.com/1270189/131036717-ba1e690b-9e6f-4415-9633-5d90cc1ce508.mp4

### Testing Instructions

Verify in each environment that:

* Expand/Collapse list is now animated
* Moving blocks is animated
* When reduce motion is requested, animations are disabled, but drag is still available.

| Site Editor |
|-----|
| <img width="1178" alt="Screen Shot 2021-08-26 at 2 24 00 PM" src="https://user-images.githubusercontent.com/1270189/131038140-4bbb4a84-ed09-4383-aa72-6ea0ab56857d.png"> |

| Post Editor |
|-----|
| <img width="1178" alt="Screen Shot 2021-08-26 at 2 26 25 PM" src="https://user-images.githubusercontent.com/1270189/131038313-1a840d10-b5ee-440f-b1cf-2ace745a89be.png"> |

| Block Navigation List | Modal |
|-----|-----|
| <img width="603" alt="Screen Shot 2021-08-26 at 2 21 49 PM" src="https://user-images.githubusercontent.com/1270189/131037848-d0405814-ec46-488a-abcf-003e1da02e1c.png"> | <img width="557" alt="Screen Shot 2021-08-26 at 2 21 56 PM" src="https://user-images.githubusercontent.com/1270189/131037862-11e7ee88-50c2-4dcc-87da-a0111a01b4fc.png"> |

### Performance

Using motion components comes with a performance cost. I did my best here to memo key items while dragging and avoid unneeded renders. When adding new blocks or removing blocks, we may see additional lag for updating items in list view. Note that some [odd laggy behavior](https://github.com/WordPress/gutenberg/issues/34310) is also present in trunk, but render costs may be more pronounced with this change. One thing to consider would be list virtualization, though I do understand that there are a11y concerns to work with. I think it'd be worth investigating in a follow up.

### UX interactions

For this first pass, folks have noted that we would like to see:

* How little styling we need
* Other enhancements like a hover drag handle to follow up PRs
* Holding drag near top/bottom to scroll will be handled in a follow up

Do let us know if any interactions don't feel quite right. Dragging so far works roughly like this:

* Straight Up/Down movement - Swap Siblings
* Left + movement - Move item after/before target item (usually under a new parent)
* Right + movement - Move item as child of target parent

### Drag And Drop E2E 

Puppeteer recently added some drag and drop E2E support, but I was unable to get this working with this drag implementation. (It currently selects text across elements instead of initiating a drag) @talldan let me know if you have any other tips. If we're not able to reasonably fix this, I'd recommend dropping the test.
